src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala

```diff
@@ -298,29 +298,23 @@ abstract class BTypes {
    * referring to BTypes.
    */
   sealed trait BType {
-    final override def toString: String = this match {
-      case UNIT   => "V"
-      case BOOL   => "Z"
-      case CHAR   => "C"
-      case BYTE   => "B"
-      case SHORT  => "S"
-      case INT    => "I"
-      case FLOAT  => "F"
-      case LONG   => "J"
-      case DOUBLE => "D"
-      case ClassBType(internalName) => "L" + internalName + ";"
-      case ArrayBType(component)    => "[" + component
-      case MethodBType(args, res)   => "(" + args.mkString + ")" + res
-    }
+
+    final override def toString: String = descriptor
 
     /**
-     * @return The Java descriptor of this type. Examples:
-     *  - int: I
-     *  - java.lang.String: Ljava/lang/String;
-     *  - int[]: [I
-     *  - Object m(String s, double d): (Ljava/lang/String;D)Ljava/lang/Object;
-     */
-    final def descriptor = toString
+      * @return The Java descriptor of this type. Examples:
+      *  - int: I
+      *  - java.lang.String: Ljava/lang/String;
+      *  - int[]: [I
+      *  - Object m(String s, double d): (Ljava/lang/String;D)Ljava/lang/Object;
+      */
+    def descriptor : String
+
+    /**
+      * generally used by descriptor to help build the descriptor. Allows several descriptors to be build from the same stringbuilder
+      * @param sb
+      */
+    private[BTypes] def buildDescriptor(sb: java.lang.StringBuilder): Unit
```

Comment:
I'd prefer to use a pattern match here instead of virtual dispatch. With an overload `override def descriptor(sb: java.lang.StringBuilder): String` in Class/Array/MethodBType this should be possible.

Reply:
> Pattern matching is a sequential operation, slowed with unapply calls

In general, you're point is valid, and in some cases the performance does make a difference (e.g. in `mapOver` #5906 ), but it is important to note that most pattern matching is on case class constructors or plain old `instanceOfs` and doesn't involve unapply calls.